src/HttpRequest.cc

```diff
@@ -668,6 +668,47 @@ HttpRequest::canHandle1xx() const
     return true;
 }
 
+bool
+HttpRequest::canUseContentLength() const
+{
+    // RFC 7230 Section 3.3.3 #4:
+    // conflicting Content-Length(s) mean a message framing error
+    if (header.conflictingContentLength())
+        return false;
+
+    // HTTP/1.0 requirements differ from HTTP/1.1
+    if (http_ver <= Http::ProtocolVersion(1,0)) {
+        const auto m = method.id();
+
+        // RFC 1945 section 8.3:
+        // "
+        //   A valid Content-Length is required on all HTTP/1.0 POST requests.
+        // "
+        if (m == Http::METHOD_POST)
+            return (content_length >= 0);
+
+        // RFC 1945 section 7.2:
+        // "
+        //   An entity body is included with a request message only when the
+        //   request method calls for one.
+        // "
+        // GET and HEAD do not define ('call for') an entity
+        if (m == Http::METHOD_GET || m == Http::METHOD_HEAD)
+            return (content_length < 0);
+
+        // other methods are not defined in RFC 1945
+    }
```

Comment:
1.0 differing from 1.1 does not preclude non-1.0 (aka "other") methods from having conditions defined in their own spec(s). RFC1945 conditions are only relevant to methods defined there AND only when used in HTTP/1.0 messages.

The fall-through is intentional to make default method handling be per the specification which Squid advertises as its version. Currently that being RFC7230.

Reply:
> 1.0 differing from 1.1 does not preclude non-1.0 (aka "other") methods from having conditions defined in their own spec(s). RFC1945 conditions are only relevant to methods defined there AND only when used in HTTP/1.0 messages.

It is technically inaccurate to call, say, the PUT or DELETE method "non-1.0" and to restrict generic RFC 1945 conditions to GET and HEAD. RFC 1945 does explicitly acknowledge the existence of other methods and even details PUT, DELETE, UNLINK, and LINK.

> The fall-through is intentional

Being intentional does not make it right. For code clarity and safety sake, I recommend following the sketch suggested in the change request, but there is another way to address my concern -- you can make the dangerous fall-through explicit. This will require a different comment to explain why we want to treat some HTTP/1.0 messages as if they were HTTP/1.1 messages _after_ handling version-agnostic requirements _and_ saying that HTTP/1.0 requirements differ from HTTP/1.1 requirements. I do not recommend this painful path, but I cannot stop you from trying to convince me that the dangerous fall-through is justifiable.

> to make default method handling be per ... RFC7230

It is not clear why we treat HTTP/1.0 GET specially, but do not treat, say, HTTP/1.0 PUT specially. HTTP/1.0 specs that we honor do not apply to just GETs and HEADs. They apply to all HTTP/1.0 messages. I see no good reason to introduce this complex fall-through logic (and the need to justify it), but if you insist on introducing it, then you have to justify it.

Reply:
Thanks for pointing out PUT, DELETE etc. I did missed the appendix.

Done in 0c94c29